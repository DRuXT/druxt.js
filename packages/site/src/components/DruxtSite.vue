<script>
import { DrupalJsonApiParams } from 'drupal-jsonapi-params'
import DruxtModule from 'druxt/dist/components/DruxtModule.vue'
import { mapActions } from 'vuex'

/**
 * The DruxtSite component renders all available Drupal block regions and
 * content, based on the specified theme.
 *
 * While Drupal provides placement configuration for blocks, it does not provide
 * any information on where each region should be placed.
 *
 * All regions are provided as scoped slots for the Druxt Wrapper component.
 *
 * @example @lang vue
 * <template>
 *   <DruxtSite theme="umami" />
 * </template>
 *
 * @example <caption>DruxtSite Wrapper component boilerplate</caption> @lang vue
 * <template>
 *   <div>
 *     <slot name="header" />
 *     <slot name="content" />
 *     <slot name="footer" />
 *   </div>
 * </template>
 *
 * <script>
 * import { DruxtSiteMixin } from 'druxt-site'
 * export default {
 *   mixins: [DruxtSiteMixin]
 * }
 *
 * @example <caption>DruxtSite with template injection</caption> @lang vue
 * <DruxtSite>
 *   <template #default="{ props, regions, theme }">
 *     <!-- Do whatever you want here -->
 *     <DruxtBlockRegion
 *       v-for="region of regions"
 *       :key="region"
 *       v-bind="props[region]"
 *     />
 *   </template>
 * </DruxtSite>
 *
 * @see {@link /api/packages/blocks/components/DruxtBlockRegion|DruxtBlockRegion}
 */
export default {
  name: 'DruxtSite',

  extends: DruxtModule,

  /** */
  props: {
    /**
     * Drupal theme ID.
     *
     * Used to filter the available regions from the Drupal Blocks JSON:API
     * resources.
     *
     * @type {string}
     */
    theme: {
      type: String,
      default: undefined,
    }
  },

  data: ({ $druxt }) => ({
    defaultTheme: ($druxt.settings.site || {}).theme,
  }),

  /** */
  computed: {
    /**
     * DruxtBlockRegion propsData for regions.
     *
     * @return {object}
     */
    props: ({ defaultTheme, regions, theme }) =>
      Object.fromEntries(regions.map((region) => [region, {
        key: region,
        name: region,
        theme: theme || defaultTheme,
      }])),

    /**
     * An array of unique region names.
     * @return {string[]}
     */
    regions: ({ model, value }) => model || value || [],
  },

  watch: {
    theme() {
      this.$fetch()
    }
  },

  methods: {
    ...mapActions({ getCollection: 'druxt/getCollection' }),
  },

  /**
   * Druxt module configuration.
   */
  druxt: {
    /**
     * Provides the available component naming options for the Druxt Wrapper.
     *
     * @param {object} context - The module component ViewModel.
     * @returns {ComponentOptions}
     */
    componentOptions: ({ defaultTheme, theme }) => [[theme || defaultTheme], ['default']],

    /**
     * Fetches theme filtered region names from the Block JSON:API resources to
     * be used to render the `<DruxtBlockRegion />`'s.
     */
    async fetchConfig() {
      const type = 'block--block'

      // If no default theme is provided, get the first available valid theme.
      if (!this.defaultTheme) {
        this.defaultTheme = await this.getCollection({
          type,
          query: new DrupalJsonApiParams()
            .addFields(type, ['theme'])
            .addFilter('plugin', 'system_main_block')
            .addPageLimit(1)
        }).then((resources) => resources.data[0].attributes.theme)
      }

      // Fetch all available regions.
      if (!this.value) {
        this.model = await this.getCollection({
          type,
          query: new DrupalJsonApiParams()
            .addFilter('theme', this.theme || this.defaultTheme)
            .addFields(type, ['region']),
        }).then((resources) => resources.data.map((resource) => resource.attributes.region).filter((v, i, s) => s.indexOf(v) === i))
      }
    },

    /**
     * Provides propsData for the DruxtWrapper.
     *
     * @param {object} context - The module component ViewModel.
     * @returns {PropsData}
     */
    propsData: ({ defaultTheme, props, regions, theme }) => ({ props, regions, theme: theme || defaultTheme }),

    /**
     * Provides the scoped slots object for the Module render function.
     *
     * A scoped slot is provided for each block region available, as per the
     * specified theme.
     *
     * Additionally, the `default` slot will render all regions, or the Nuxt
     * component when no block region data is available.
     *
     * @example <caption>DruxtSite**Theme**.vue</caption> @lang vue
     * <template>
     *   <div>
     *     <slot name="content" />
     *     <slot :name="region_name" />
     *   </div>
     * </template>
     *
     * @return {ScopedSlots} The Scoped slots object.
     */
    slots(h) {
      // If no regions, return Nuxt component.
      if (!this.regions.length) {
        return { default: () => h('Nuxt') }
      }

      // Build scoped slots for each region.
      const scopedSlots = {
        ...Object.fromEntries(this.regions.map((region) => [region, (attrs) => h('DruxtBlockRegion', {
          attrs,
          key: region,
          props: this.props[region],
          ref: region,
        })]))
      }

      // Build default slot.
      scopedSlots.default = (attrs) => this.regions.map((region) => (scopedSlots[region] || (() => {}))(attrs))

      return scopedSlots
    },

    /**
     * Druxt development template tool configuration.
     */
    template: {
      debug: '{ props, regions, theme }',
    }
  }
}

/**
 * Provides the available naming options for the Wrapper component.
 *
 * @typedef {array[]} ComponentOptions
 *
 * @example @lang js
 * [
 *   'DruxtSite[Theme][Langcode]',
 *   'DruxtSite[Theme]',
 *   'DruxtSite[Default][Langcode]',
 *   'DruxtSite[Default]',
 * ]
 *
 * @example <caption>Umami</caption> @lang js
 * [
 *   'DruxtSiteUmamiEn',
 *   'DruxtSiteUmami',
 *   'DruxtSiteDefaultEn',
 *   'DruxtSiteDefault',
 * ]
 */

/**
 * Provides propsData for use in the Wrapper component.
 *
 * @typedef {object} PropsData
 * @param {object} props - DruxtBlockRegion propsData for regions.
 * @param {string[]} regions - An array of unique region names.
 * @param {string} theme - Drupal theme ID.
 *
 * @example @lang js
 * {
 *   props: {
 *     content: {
 *       name: 'content',
 *       theme: 'umami',
 *     },
 *     ...
 *   },
 *   regions: ['breadcrumbs', 'header', 'content', ...],
 *   theme: 'umami',
 * }
 */

/**
 * Provides scoped slots for use in the Wrapper component.
 *
 * @typedef {object} ScopedSlots
 * @param {function} * - Slot per region.
 * @param {function} default - All regions.
 *
 * @example <caption>DruxtSite**Theme**.vue</caption> @lang vue
 * <template>
 *   <div>
 *     <slot name="content" />
 *     <slot :name="region_name" />
 *   </div>
 * </template>
 */
</script>
