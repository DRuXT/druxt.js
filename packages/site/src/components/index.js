/**
 * The DruxtSite Vue.js component.
 *
 * Renders all available block regions and content based on Drupals
 * configuration and routing information.
 *
 * @type {object}
 * @exports DruxtSite
 * @name DruxtSite
 * @see {@link ./components/DruxtSite|DruxtSite}
 * @example @lang vue
 * <template>
 *   <DruxtSite theme="umami" />
 * </template>
 *
 * <script>
 *   import { DruxtSite } from 'druxt-site'
 *
 *   export default {
 *     components: { DruxtSite }
 *   }
 * </script>
 */
 export { default as DruxtSite } from './DruxtSite.vue'
