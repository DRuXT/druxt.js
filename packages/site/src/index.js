/**
 * Vue.js Mixin.
 *
 * Registers props for use by Druxt slot theme components.
 *
 * @type {object}
 * @exports DruxtSiteMixin
 * @see {@link ./mixins/site|DruxtSiteMixin}
 * @example @lang vue
 * <template>
 *   <div>
 *     <slot v-for="region of regions" :key="region" :name="region" />
 *   </div>
 * </template>
 *
 * <script>
 * import { DruxtSiteMixin } from 'druxt-site'
 *
 * export default {
 *   mixins: [DruxtSiteMixin],
 * }
 * </script>
 */
export { DruxtSiteMixin } from './mixins/site'

/**
 * Default function to alert user to incorrectly installed module.
 *
 * This was added as part of the @nuxt/kit update due to breaking changes.
 */
export default () => {
  throw new Error("DruxtSite Nuxt module must be installed as 'druxt-site/nuxt'")
}
