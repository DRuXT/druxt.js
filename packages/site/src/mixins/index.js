/**
 * Vue.js Mixin.
 *
 * Registers props for use by Druxt slot theme components.
 *
 * @type {object}
 * @exports DruxtSiteMixin
 * @see {@link ./mixins/site|DruxtSiteMixin}
 * @example @lang vue
 * <template>
 *   <div>
 *     <slot v-for="region of regions" :key="region" :name="region" />
 *   </div>
 * </template>
 *
 * <script>
 * import { DruxtSiteMixin } from 'druxt-site'
 *
 * export default {
 *   mixins: [DruxtSiteMixin],
 * }
 * </script>
 */
 export { DruxtSiteMixin } from './site'
