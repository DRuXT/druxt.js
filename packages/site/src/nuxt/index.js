import { addLayout, defineNuxtModule, installModule } from '@nuxt/kit'
import { existsSync } from 'fs'
import { resolve } from 'path'
// import DruxtSiteStorybook from '../nuxtStorybook'

/**
 * Nuxt module function to install Druxt Site.
 *
 * - Adds the DruxtSite component.
 * - Adds the core modules for DruxtJS Site.
 * - Adds default configuration for @nuxtjs/proxy.
 * - Enables Vuex store.
 *
 * @param {ModuleOptions} moduleOptions - The Nuxt.js module options.
 */
const DruxtSiteNuxtModule = defineNuxtModule({
  meta: {
    name: 'druxt-site',
  },
  defaults: {
    baseUrl: '',
    endpoint: '/jsonapi'
  },

  async setup(moduleOptions, nuxt) {
    // Prevent issue "FATAL: Cannot determine nuxt version! Is current
    // instance passed?".
    nuxt._version = nuxt._version || '2.'
    // This is required to prevent "FATAL: nuxt.options._layers is not iterable"
    // error when using `installModule()`.
    nuxt.options._layers = nuxt.options._layers || []

    // Set default options.
    const options = {
      baseUrl: moduleOptions.baseUrl,
      ...nuxt.options?.druxt,
      proxy: {
        api: false,
        files: true,
        ...nuxt.options?.druxt?.proxy,
      },
      site: {
        layout: true,
        ...nuxt.options?.druxt?.site,
        ...moduleOptions,
      },
    }
    nuxt.options.druxt = options

    // Register components directories.
    nuxt.hook('components:dirs', dirs => {
      dirs.push({ path: resolve(__dirname, '../dist/components') })
    })

    // Add Druxt modules.
    const druxtModules = [
      'druxt/nuxt',
      'druxt-blocks/nuxt',
      'druxt-breadcrumb/nuxt',
      'druxt-entity/nuxt',
      'druxt-menu/nuxt',
      'druxt-router/nuxt',
      'druxt-schema/nuxt',
      'druxt-views/nuxt'
    ]
    for (const module of druxtModules) {
      await installModule(module, {}, nuxt)
    }

    // Add default layout.
    if (!(await existsSync(resolve(nuxt.options.srcDir, nuxt.options.dir.layouts))) && options.site.layout) {
      addLayout(resolve(__dirname, '../dist/layouts/default.vue'), 'default')
    }

    // Nuxt Storybook.
    // @TODO - @nuxt/kit and @nuxt/storybook aren't compatible.
    // this.nuxt.hook('storybook:config', async ({ stories }) => {
    //   await DruxtSiteStorybook.call(this, { options, stories })
    // })
  }
})

export default DruxtSiteNuxtModule

/**
 * Module options object.
 *
 * @typedef {object} ModuleOptions
 * @see {@link ./typedefs/moduleOptions|ModuleOptions}
 */
