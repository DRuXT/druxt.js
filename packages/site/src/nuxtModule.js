import { resolve } from 'path'

/**
 * Nuxt module function to install Druxt Site.
 *
 * - Adds the DruxtSite component.
 * - Adds the core modules for DruxtJS Site.
 * - Adds default configuration for @nuxtjs/proxy.
 * - Enables Vuex store.
 *
 * @param {ModuleOptions} moduleOptions - The Nuxt.js module options.
 */
const DruxtSiteNuxtModule = function (moduleOptions = {}) {
  // Use root level Druxt options.
  if (typeof this.options === 'undefined' || !this.options.druxt) {
    throw new TypeError('Druxt settings missing.')
  }

  // Add plugin.
  this.addPlugin({
    src: resolve(__dirname, '../nuxt/plugin.js'),
    fileName: 'druxt-site.js',
    options: this.options.druxt
  })

  // Add Nuxt.js modules.
  const modules = [
    '@nuxtjs/proxy',
    'druxt',
    'druxt-blocks',
    'druxt-breadcrumb',
    'druxt-entity',
    'druxt-menu',
    'druxt-router',
    'druxt-schema',
    'druxt-views'
  ]
  for (const key in modules) {
    this.addModule(modules[key])
  }

  // Setup proxy for 'sites/default/files'.
  // @todo {@link https://github.com/nuxt-community/proxy-module/issues/80|Suppress warning}
  if (typeof this.options.proxy === 'undefined') {
    this.options.proxy = [this.options.druxt.baseUrl + '/sites/default/files']
  }

  // Enable JSON:API Menu items by default.
  if (typeof ((this.options.druxt || {}).menu || {}).jsonApiMenuItems === 'undefined') {
    this.options.druxt.menu = {
      ...this.options.druxt.menu,
      ...{ jsonApiMenuItems: true }
    }
  }

  // Enable Vuex Store.
  this.options.store = true
}

DruxtSiteNuxtModule.meta = require('../package.json')

export { DruxtSiteNuxtModule }

/**
 * Module options object.
 *
 * @typedef {object} ModuleOptions
 * @see {@link ./typedefs/moduleOptions|ModuleOptions}
 */
