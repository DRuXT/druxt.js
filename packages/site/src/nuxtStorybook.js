import { resolve } from 'path'
import { DrupalJsonApiParams } from 'drupal-jsonapi-params'
import { DruxtClient } from 'druxt'

const titleFn = (parts) =>
  parts.map((part) =>
    part.charAt(0).toUpperCase() + part.slice(1).replace(/_/g, ' ')
  ).join('/')

export default async function ({ options, stories }) {
  const { addTemplate } = this

  const druxt = new DruxtClient(options.baseUrl, { ...options, proxy: { api: false } })

  const type = 'block--block'
  const query = new DrupalJsonApiParams().addFields(type, ['theme'])
  const resources = (await druxt.getCollectionAll(type, query)).map((collection) => collection.data).flat()
  const themes = [...new Set(resources.map((o) => o.attributes.theme))].sort()
  const siteStories = themes.map((theme) => {
    addTemplate({
      src: resolve(__dirname, '../templates/druxt-site.stories.js'),
      fileName: `stories/druxt-site.${theme}.stories.js`,
      options: {
        theme,
        title: titleFn(['Druxt', 'Site', 'Themes', theme])
      }
    })
    return resolve(this.options.buildDir, `./stories/druxt-site.${theme}.stories.js`)
  })
  stories.push.apply(stories, siteStories)
}
