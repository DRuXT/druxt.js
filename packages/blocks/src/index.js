import { DruxtBlocksNuxtModule } from './nuxtModule'

/**
 * The Nuxt.js module function.
 *
 * Installs the module functionality in a Nuxt application.
 *
 * @type {Function}
 * @exports default
 * @name DruxtBlocksModule
 * @see {@link ./nuxtModule|DruxtBlocksModule}
 *
 * @example <caption>nuxt.config.js</caption> @lang js
 * module.exports = {
 *   modules: [
 *     'druxt-blocks'
 *   ],
 *   druxt: {
 *     baseUrl: 'https://demo-api.druxtjs.org'
 *   }
 * }
 */
export default DruxtBlocksNuxtModule

/**
 * Vue.js Mixin.
 *
 * @type {object}
 * @exports DruxtBlocksBlockMixin
 * @see {@link ./mixins/block|DruxtBlocksBlockMixin}
 */
 export { DruxtBlocksBlockMixin } from './mixins/block'
