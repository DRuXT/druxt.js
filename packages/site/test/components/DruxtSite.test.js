import 'regenerator-runtime/runtime'
import { createLocalVue, mount } from '@vue/test-utils'
import Vuex from 'vuex'

import { DruxtClient, DruxtStore } from '../../../druxt/src'
import DruxtSite from '../../src/components/DruxtSite.vue'

jest.mock('axios')

// Setup local vue instance.
const localVue = createLocalVue()
localVue.use(Vuex)

let store

const mountComponent = (pending, options) => {
  const mocks = {
    $druxt: { settings: {} },
    $fetchState: { pending }
  }
  const propsData = { theme: 'umami' }
  const stubs = ['DruxtBlockRegion', 'Nuxt']

  return mount(DruxtSite, { localVue, mocks, propsData, store, stubs, ...options })
}

describe('DruxtSite component', () => {
  beforeEach(() => {
    // Setup vuex store.
    store = new Vuex.Store()

    DruxtStore({ store })
    store.$druxt = new DruxtClient('https://demo-api.druxtjs.org')

    store.app = { context: { error: jest.fn() }, store }
  })

  test('defaults', async () => {
    const wrapper = mountComponent(false)
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    expect(wrapper.vm.theme).toBe('umami')
    expect(wrapper.vm.regions).toStrictEqual([
      'breadcrumbs',
      'highlighted',
      'pre_header',
      'banner_top',
      'header',
      'content',
      'bottom',
      'footer',
      'tabs',
      'page_title',
      'content_bottom'
    ])
    expect(Object.keys(wrapper.vm.$refs)).toStrictEqual(['component', ...wrapper.vm.regions])

    // Druxt Component mixin.
    expect(wrapper.vm.component.is).toBe('DruxtWrapper')
    expect(wrapper.vm.component.options).toStrictEqual(['DruxtSiteUmami', 'DruxtSiteDefault'])

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('pending', async () => {
    const wrapper = mountComponent(true)

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('default slot', async () => {
    const scopedSlots = { default: jest.fn() }
    const wrapper = mountComponent(false, { scopedSlots })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    wrapper.vm.getScopedSlots().default()
    expect(scopedSlots.default).toHaveBeenCalledWith({
      props: wrapper.vm.props,
      regions: wrapper.vm.regions,
      theme: wrapper.vm.theme
    })
  })

  test('no blocks', async () => {
    const wrapper = mountComponent(false)
    expect(wrapper.vm.regions.length).toBe(0)
    expect(wrapper.vm.getScopedSlots().default().componentOptions.tag).toBe('Nuxt')
  })
})
