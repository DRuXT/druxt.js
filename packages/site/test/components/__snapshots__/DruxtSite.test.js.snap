// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DruxtSite component defaults 1`] = `
"<div>
  <div props=\\"[object Object]\\" regions=\\"breadcrumbs,highlighted,pre_header,banner_top,header,content,bottom,footer,tabs,page_title,content_bottom\\" theme=\\"umami\\">
    <druxtblockregion-stub></druxtblockregion-stub>
    <druxtblockregion-stub></druxtblockregion-stub>
    <druxtblockregion-stub></druxtblockregion-stub>
    <druxtblockregion-stub></druxtblockregion-stub>
    <druxtblockregion-stub></druxtblockregion-stub>
    <druxtblockregion-stub></druxtblockregion-stub>
    <druxtblockregion-stub></druxtblockregion-stub>
    <druxtblockregion-stub></druxtblockregion-stub>
    <druxtblockregion-stub></druxtblockregion-stub>
    <druxtblockregion-stub></druxtblockregion-stub>
    <druxtblockregion-stub></druxtblockregion-stub>
  </div>
</div>"
`;

exports[`DruxtSite component pending 1`] = `"<div></div>"`;
